bls/algebra/src/bls/keys.rs

```diff
@@ -177,6 +223,22 @@ impl PublicKey {
     }
 }
 
+impl PartialEq for PublicKey {
+    fn eq(&self, other: &Self) -> bool {
+        // The equality operator in G2Projective is very slow
+        self.pk.x == other.pk.x && self.pk.y == other.pk.y && self.pk.z == other.pk.z
+    }
+}
+
+impl Hash for PublicKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // Only hash based on `y` for slight speed improvement
```

Comment:
Why is this okay

Reply:
Take a look at the requirements for hash/equal described in the [docs for HashSet](https://doc.rust-lang.org/std/collections/struct.HashSet.html):

`k1 == k2` must imply  `hash(k1) == hash(k2)`

While the reverse is not true.  (`hash(k1) == hash(k2)` need not imply `k1 == k2`)


`hash(k1)` is being used to decide which hash bucket (out of hundreds) a key will land in the HashSet.  A collision means the equal operator will be used to figure out if they are the same key.

Reply:
Got it